pkg/controller/replicaset/replica_set.go

```diff
@@ -596,11 +590,19 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 		}
 	}
 
-	canAdoptFunc := getCanAdoptFunc(rsc, rs)
-	cm := controller.NewPodControllerRefManager(rsc.podControl, rs, selector, controllerKind, canAdoptFunc)
-	// NOTE: filteredPods are pointing to objects from cache - if you need to
-	// modify them, you need to copy it first.
-	filteredPods, err = cm.ClaimPods(filteredPods)
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing Pods (see #42639).
+	canAdoptFunc := controller.RecheckDeletionTimestamp(func() (metav1.Object, error) {
+		fresh, err := rsc.kubeClient.ExtensionsV1beta1().ReplicaSets(rs.Namespace).Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if fresh.UID != rs.UID {
+			return nil, fmt.Errorf("original ReplicaSet %v/%v is gone: got uid %v, wanted %v", rs.Namespace, rs.Name, fresh.UID, rs.UID)
+		}
+		return fresh, nil
+	})
+	filteredPods, err = rsc.claimPods(rs, selector, canAdoptFunc, filteredPods)
```

Comment:
No unit test is added for `rsc.claimPods` as it is covered by integration and e2e tests.